### PR TITLE
rg-134: Template create and edit

### PR DESCRIPTION
**Pull request by honkster-nazarii**
Demo:

https://github.com/BinaryStudioAcademy/bsa-winter-2023-2024-resumegemm/assets/133093930/f8920305-b735-4cda-aeb9-f1c70083f10b

### CODE REVIEW

File: backend/src/bundles/templates/template.controller.ts [hunk 1]
```diff
@@ -89,7 +93,7 @@ class TemplateController extends Controller {
         if (!template) {
             throw new HTTPError({
                 status: HttpCode.BAD_REQUEST,
-                message: 'User with this id not found',
+                message: 'Template with this id not found',
```
Comment (MaxPoplavskiy) [inline backend/src/bundles/templates/template.controller.ts:96]: consider moving to enum

File: frontend/src/bundles/edit-template/components/template-item/template-item.tsx [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    name,
+    type,
+    content,
+    styles,
+}) => {
+    return (
+        <div key={id} style={styles}>
+            {type === 'h1' && <h1>{content}</h1>}
+            {type === 'p' && <p>{content}</p>}
+            {type === 'img' && <img src={content} alt={name} />}
```
Comment (MaxPoplavskiy) [inline frontend/src/bundles/edit-template/components/template-item/template-item.tsx:16]: consider moving to a callback

File: frontend/src/bundles/home/pages/templates.tsx [hunk 3]
```diff
@@ -1,19 +1,41 @@
...
+    const navigate = useNavigate();
+
+    const handleCreateTemplate = useCallback(() => {
+        dispatch(createTemplate())
+            .unwrap()
+            .then((newTemplate: TemplateDto) => {
+                navigate(`${AppRoute.TEMPLATE_EDITOR}/${newTemplate.id}`);
+            })
+            .catch(() => {
+                toast.error('Failed to create template');
```
Comment (MaxPoplavskiy) [inline frontend/src/bundles/home/pages/templates.tsx:36]: consider moving to enum

**Review by MaxPoplavskiy — CHANGES_REQUESTED** [review]

**Review by MaxPoplavskiy — APPROVED** [review]

**Review by Valentyn13 — APPROVED** [review]

**Comment by honkster-nazarii** [comment]
[Template creating and basic editing](https://trello.com/c/mONFhwm6/134-template-creating-and-basic-editing)

File: frontend/src/bundles/common/components/checkbox/checkbox.tsx [hunk 4]
```diff
@@ -12,7 +12,7 @@ type Properties = {
 
 const Checkbox: React.FC<Properties> = ({
     label,
-    checked = false,
+    checked,
```
Comment (Vitaliy-Kalenichenko) [inline frontend/src/bundles/common/components/checkbox/checkbox.tsx:15]: Why it's been changed?

File: frontend/src/bundles/edit-template/components/template-block/template-block.tsx [hunk 5]
```diff
@@ -0,0 +1,17 @@
+import { type Block } from '~/bundles/templates-page/types/types';
+
+import { TemplateItem } from '../template-item/template-item';
+
+type Properties = Block;
+
+const TemplateBlock: React.FC<Properties> = ({ id, items, styles }) => {
+    return (
+        <div key={id} style={styles}>
```
Comment (Vitaliy-Kalenichenko) [inline frontend/src/bundles/edit-template/components/template-block/template-block.tsx:9]: Do you really need `key` here?

File: frontend/src/bundles/edit-template/store/actions.ts [hunk 6]
```diff
@@ -1,39 +1,35 @@
...
 import { name as sliceName } from './slice.js';
 
-type EditTemplayePayload = Partial<TemplateUpdateItemRequestDto> & {
-    id: string;
-};
+const getTemplateById = createAsyncThunk<TemplateDto, string, AsyncThunkConfig>(
+    `${sliceName}/get-by-id`,
+    async (id, { extra }) => {
+        const { templateApi } = extra;
+        return await templateApi.getTemplateById(id);
```
Comment (Vitaliy-Kalenichenko) [inline frontend/src/bundles/edit-template/store/actions.ts:12]:
```suggestion
        return templateApi.getTemplateById(id);
```

File: frontend/src/bundles/common/components/checkbox/checkbox.tsx (hunk 4 above)
Comment (honkster-nazarii) [reply frontend/src/bundles/common/components/checkbox/checkbox.tsx:15]: I changed defaultChecked to checked, and `checked = false` would make checkbox to be only false if checked prop is not provided

File: frontend/src/bundles/edit-template/components/template-container/template-container.tsx [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+import { TemplateBlock } from '../template-block/template-block';
+
+type Properties = Container;
+const TemplateContainer: React.FC<Properties> = ({ id, blocks, styles }) => {
+    return (
+        <div key={id} style={styles}>
+            {blocks.map((block) => {
+                if (block.enabled) {
+                    return <TemplateBlock key={block.id} {...block} />;
+                }
```
Comment (Vitaliy-Kalenichenko) [inline frontend/src/bundles/edit-template/components/template-container/template-container.tsx:12]:
```suggestion
                block.enabled && <TemplateBlock key={block.id} {...block} />;
```

**Review by Vitaliy-Kalenichenko — CHANGES_REQUESTED** [review]

File: frontend/src/bundles/edit-template/components/template-block/template-block.tsx [hunk 8]
```diff
@@ -0,0 +1,17 @@
+import { type Block } from '~/bundles/templates-page/types/types';
```
Comment (Vitaliy-Kalenichenko) [inline frontend/src/bundles/edit-template/components/template-block/template-block.tsx:1]: Try to come up with more descriptive name

File: frontend/src/bundles/edit-template/components/template-item/template-item.tsx [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+const TemplateItem: React.FC<Properties> = ({
+    id,
+    name,
+    type,
+    content,
+    styles,
+}) => {
+    const handleType = useCallback(
+        (type: string) => {
+            switch (type) {
```
Comment (Vitaliy-Kalenichenko) [inline frontend/src/bundles/edit-template/components/template-item/template-item.tsx:16]:
```suggestion
            switch (tagName) {
```

File: frontend/src/bundles/edit-template/components/template-item/template-item.tsx [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+                case 'h1': {
```
Comment (Vitaliy-Kalenichenko) [inline frontend/src/bundles/edit-template/components/template-item/template-item.tsx:17]: Consider creating enum for tag names

File: frontend/src/bundles/edit-template/store/actions.ts (hunk 6 above)
Comment (honkster-nazarii) [reply frontend/src/bundles/edit-template/store/actions.ts:12]: async func should have await statement

Comment (apofis96) [reply frontend/src/bundles/edit-template/store/actions.ts:12]: Why it should be async than?

Comment (honkster-nazarii) [reply frontend/src/bundles/edit-template/store/actions.ts:12]: because its a async fetch request

Comment (Icerock10) [reply frontend/src/bundles/edit-template/store/actions.ts:12]: Outside of try/catch blocks, return await is redundant.

**Review by Vitaliy-Kalenichenko — APPROVED** [review]

**Pull request merged**